lib/handlers/pools/provider-migration/v3/traffic-switch-v3-pool-provider.ts

```diff
@@ -0,0 +1,120 @@
+import { IV3PoolProvider, log, metric, MetricLoggerUnit, V3PoolAccessor } from '@uniswap/smart-order-router'
+import { Token } from '@uniswap/sdk-core'
+import { FeeAmount, Pool } from '@uniswap/v3-sdk'
+import { ProviderConfig } from '@uniswap/smart-order-router/build/main/providers/provider'
+import { POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION } from './pool-provider-traffic-switch-configuration'
+
+export type TrafficSwitchPoolProviderProps = {
+  currentPoolProvider: IV3PoolProvider
+  targetPoolProvider: IV3PoolProvider
+  sourceOfTruthPoolProvider: IV3PoolProvider
+}
+
+export class TrafficSwitchV3PoolProvider implements IV3PoolProvider {
+  private readonly currentPoolProvider: IV3PoolProvider
+  private readonly targetPoolProvider: IV3PoolProvider
+  private readonly sourceOfTruthPoolProvider: IV3PoolProvider
+
+  protected readonly SHOULD_SWITCH_TRAFFIC = () =>
+    POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.switchPercentage > this.getRandomPercentage()
+  protected readonly SHOULD_SAMPLE_TRAFFIC = () =>
+    POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.samplingPercentage > this.getRandomPercentage()
+
+  constructor({ currentPoolProvider, targetPoolProvider, sourceOfTruthPoolProvider }: TrafficSwitchPoolProviderProps) {
+    this.currentPoolProvider = currentPoolProvider
+    this.targetPoolProvider = targetPoolProvider
+    this.sourceOfTruthPoolProvider = sourceOfTruthPoolProvider
+  }
+
+  getPoolAddress(
+    tokenA: Token,
+    tokenB: Token,
+    feeAmount: FeeAmount
+  ): {
+    poolAddress: string
+    token0: Token
+    token1: Token
+  } {
+    // The underlying logic for getting the pool address is always from the
+    // source of truth pool provider, and the result is deterministic.
+    // It doesn't matter which pool provider we use, so we can start with the target
+    // pool provider, which will delegate to the source of truth pool provider.
+    return this.targetPoolProvider.getPoolAddress(tokenA, tokenB, feeAmount)
+  }
+
+  async getPools(tokenPairs: [Token, Token, FeeAmount][], providerConfig?: ProviderConfig): Promise<V3PoolAccessor> {
+    const currentProviderPools = await this.currentPoolProvider.getPools(tokenPairs, providerConfig)
+
+    metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_TOTAL', 1, MetricLoggerUnit.None)
+    const sampleTraffic = this.SHOULD_SAMPLE_TRAFFIC()
+    if (sampleTraffic) {
+      metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_SAMPLING', 1, MetricLoggerUnit.None)
+      const targetProviderPools = await this.targetPoolProvider.getPools(tokenPairs, providerConfig)
+
+      // If we need to sample the traffic, we don't want to make it a blocking I/O
```

Comment:
if we want to swtich traffic, then this SHOULD be a blocking call right?

Reply:
i guess you are awaiting at line 64 but you are calculating the same thing twice it seems?

Reply:
Modified the call structure, so that the call to currentProviderPools & targetProviderPools will be at most once respectively.